chain/indexer/src/snapshot/validate.rs

```diff
@@ -0,0 +1,219 @@
+        if !missing_tables.is_empty() {
+            return Err(SnapshotError::Validation(format!(
+                "Missing required table(s): {}. Available tables: [{}]",
+                missing_tables.join(", "),
+                tables.join(", ")
+            )));
+        }
+
+        // Get snapshot metadata using the log table (block_number stored as 8-byte big-endian blob)
+        let log_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM log")
+            .fetch_one(&mut conn)
+            .await
+            .map_err(|e| SnapshotError::Validation(format!("Cannot count logs: {e}")))?;
+
+        // Get latest block number from log table (handling blob format)
+        let latest_block: Option<i64> = match sqlx::query_scalar::<_, Vec<u8>>("SELECT MAX(block_number) FROM log")
+            .fetch_optional(&mut conn)
+            .await
+            .map_err(|e| SnapshotError::Validation(format!("Cannot get latest block: {e}")))?
+        {
+            Some(blob) if blob.len() == 8 => {
+                // Convert 8-byte big-endian blob to i64
+                let bytes: [u8; 8] = blob
+                    .try_into()
+                    .map_err(|_| SnapshotError::Validation("Invalid block_number blob length".to_string()))?;
+                Some(i64::from_be_bytes(bytes))
+            }
+            Some(_) => {
+                return Err(SnapshotError::Validation(
+                    "Invalid block_number blob format".to_string(),
+                ));
+            }
+            None => None,
+        };
+
+        // Validate that we have some data
+        if log_count == 0 {
+            warn!("Snapshot database contains no logs");
+        }
+
+        Ok(SnapshotInfo {
+            log_count: log_count as u64,
+            latest_block: latest_block.map(|b| b as u64),
+            tables: tables.len(),
+            sqlite_version,
+            db_size,
+        })
+    }
+
+    /// Checks data consistency
+    pub async fn check_data_consistency(&self, db_path: &Path) -> SnapshotResult<()> {
+        let options = SqliteConnectOptions::new().filename(db_path).read_only(true);
+
+        let mut conn = SqliteConnection::connect_with(&options)
+            .await
+            .map_err(|e| SnapshotError::Validation(format!("Cannot open database: {e}")))?;
+
+        // Check for NULL block_numbers (blob columns can't be negative in the same way)
+        let invalid_logs: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM log WHERE block_number IS NULL")
+            .fetch_one(&mut conn)
+            .await
+            .map_err(|e| SnapshotError::Validation(format!("Cannot check log consistency: {e}")))?;
+
+        if invalid_logs > 0 {
+            return Err(SnapshotError::Validation(format!(
+                "Found {invalid_logs} logs with invalid block numbers",
+            )));
+        }
+
+        Ok(())
+    }
+}
+
+impl Default for SnapshotValidator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
This could be derived by the macro probably.

Reply:
Yes, you can do `#[derive(Default)]` if all the members have default impls.